src/user-management/user-management.ts

```diff
@@ -294,6 +314,158 @@ export class UserManagement {
+    cookiePassword = process.env.WORKOS_COOKIE_PASSWORD,
+  }: SessionCookieOptions): Promise<
+    | AuthenticateWithSessionCookieSuccessResponse
+    | AuthenticateWithSessionCookieFailedResponse
+  > {
+    if (!cookiePassword) {
+      throw new Error('Cookie password is required');
+    }
+
+    if (!this.jwks) {
+      throw new Error('Must provide clientId to initialize JWKS');
+    }
+
+    if (!sessionCookie) {
+      return { authenticated: false, reason: 'no_session_cookie_provided' };
+    }
+
+    const session = await unsealData<SessionCookieData>(sessionCookie, {
+      password: cookiePassword,
+    });
+
+    if (!session.accessToken) {
+      return { authenticated: false, reason: 'invalid_session_cookie' };
+    }
+
+    if (!(await this.isValidJwt(session.accessToken))) {
+      return { authenticated: false, reason: 'invalid_jwt' };
+    }
+
+    const {
+      sid: sessionId,
+      org_id: organizationId,
+      role,
+      permissions,
+    } = decodeJwt<AccessToken>(session.accessToken);
+
+    return {
+      authenticated: true,
+      sessionId,
+      organizationId,
+      role,
+      permissions,
+    };
+  }
+
+  private async isValidJwt(accessToken: string): Promise<boolean> {
+    if (!this.jwks) {
+      throw new Error('Must provide clientId to initialize JWKS');
+    }
+
+    try {
+      await jwtVerify(accessToken, this.jwks);
+      return true;
+    } catch (e) {
+      return false;
+    }
+  }
+
+  async refreshAndSealSessionData({
+    sessionCookie,
+    cookiePassword = process.env.WORKOS_COOKIE_PASSWORD,
+  }: SessionCookieOptions): Promise<RefreshAndSealSessionDataResponse> {
+    if (!cookiePassword) {
+      throw new Error('Cookie password is required');
+    }
+
+    if (!sessionCookie) {
+      return { authenticated: false, reason: 'no_session_cookie_provided' };
+    }
+
+    const session = await unsealData<SessionCookieData>(sessionCookie, {
+      password: cookiePassword,
+    });
+
+    if (!session.refreshToken || !session.user) {
+      return { authenticated: false, reason: 'invalid_session_cookie' };
+    }
+
+    const { accessToken, refreshToken } =
+      await this.authenticateWithRefreshToken({
+        clientId: this.workos.clientId as string,
+        refreshToken: session.refreshToken,
+      });
+
+    // Refresh tokens are single use, so update the session with the
+    // new access and refresh tokens
+    const encryptedSession = await sealData(
+      {
+        accessToken,
+        refreshToken,
+        user: session.user,
+        impersonator: session.impersonator,
+      },
+      { password: cookiePassword },
+    );
+
+    return { authenticated: true, sealedSessionData: encryptedSession };
+  }
+
+  async authenticateWithCodeAndSealSessionData({
+    code,
+    cookiePassword = process.env.WORKOS_COOKIE_PASSWORD,
+  }: { code: string } & Pick<
+    SessionCookieOptions,
+    'cookiePassword'
+  >): Promise<AuthenticateWithCodeAndSealSessionDataResponse> {
```

Comment:
This is a new method to authenticate with a code and seal the session data. Same questions as above:
* How do we feel about the overall interface?
* Do we feel it provides a good level of abstraction?
* Is the naming clear?

Separately, I'm wondering if it might be better to provide a general `sealSessionData` method that takes the return value of all of the `authenticate*` methods. Alternatively it'd be cool if we had some sort of builder pattern for this process so devs can handle as much or as little of the process. For example, they could call `authenticateWithCode().session()` to get the unsealed session data, or `authenticateWithCode().sealSession()` to let us handle the data sealing.

Reply:
Yeah I like that, will add.